tortoisebeacon/tortoisebeacon.go

```
@@ -0,0 +1,450 @@
+	lateProposals   map[types.Hash32][]types.ATXID
+
+	votesMu sync.RWMutex
+	votes   map[EpochRoundPair][]types.Hash32
+
+	beaconsMu sync.RWMutex
+	beacons   map[types.EpochID]types.Hash32
+	// beaconsCh's value has channel type to be able to implement getting a beacon value in a blocking way,
+	// i.e. read from the channel
+	beaconsCh map[types.EpochID]chan types.Hash32
+}
+
+func New(conf Config, messageReceiver messageReceiver, messageSender messageSender, beaconCalculator beaconCalculator, atxDB *activation.DB, layerTicker chan types.LayerID, logger log.Log) *TortoiseBeacon {
+	return &TortoiseBeacon{
+		config:             conf,
+		Log:                logger,
+		closed:             make(chan struct{}),
+		messageReceiver:    messageReceiver,
+		messageSender:      messageSender,
+		beaconCalculator:   beaconCalculator,
+		atxDB:              atxDB,
+		layerTicker:        layerTicker,
+		networkDelta:       time.Duration(conf.WakeupDelta) * time.Second,
+		currentRounds:      make(map[types.EpochID]int),
+		timelyProposalsMap: make(map[types.Hash32][]types.ATXID),
+		delayedProposals:   make(map[types.Hash32][]types.ATXID),
+		lateProposals:      make(map[types.Hash32][]types.ATXID),
+	}
+}
+
+// Start starts listening for layers and outputs.
+func (tb *TortoiseBeacon) Start() error {
+	tb.Log.Info("Starting %v", protoName)
+
+	go tb.listenLayers()
+
+	go func() {
+		if err := tb.listenMessages(); err != nil {
+			// TODO(nkryuchkov): handle error
+			return
+		}
+	}()
+
+	go tb.cleanupLoop()
+
+	return nil
+}
+
+func (tb *TortoiseBeacon) Close() error {
+	tb.Log.Info("Closing %v", protoName)
+	close(tb.closed)
+
+	return nil
+}
+
+func (tb *TortoiseBeacon) Get(epochID types.EpochID) (types.Hash32, bool) {
+	tb.beaconsMu.RLock()
+	beacon, ok := tb.beacons[epochID]
+	tb.beaconsMu.RUnlock()
+
+	return beacon, ok
+}
+
+func (tb *TortoiseBeacon) WaitAndGet(epochID types.EpochID) types.Hash32 {
+	tb.beaconsMu.RLock()
+	ch := tb.beaconsCh[epochID]
```

what happens if the key is not present?

---

this also means that only one caller can WaitAndGet right?

---

@antonlerner I need to return an error If the key is not present. Right, only one caller can `WaitAndGet`. Should many be able to call it?